src/librustc_mir/hair/cx/expr.rs

```diff
@@ -589,12 +589,87 @@ fn make_mirror_unadjusted<'a, 'gcx, 'tcx>(cx: &mut Cx<'a, 'gcx, 'tcx>,
+                    cx
+                        .tables()
+                        .node_id_to_type(source.hir_id)
+                        .ty_adt_def()
+                        .and_then(|adt_def| {
+                        match def {
+                            Def::VariantCtor(variant_id, CtorKind::Const) => {
+                                let idx = adt_def.variant_index_with_id(variant_id);
+                                let (d, o) = adt_def.discriminant_def_for_variant(idx);
+                                use rustc::ty::util::IntTypeExt;
+                                let ty = adt_def.repr.discr_type().to_ty(cx.tcx());
+                                Some((d, o, ty))
+                            }
+                            _ => None,
+                        }
+                    })
+                } else {
+                    None
+                };
+                if let Some((did, offset, ty)) = var {
+                    let mk_const = |tcx: ty::TyCtxt<'a, 'gcx, 'tcx>, val| Expr {
+                        temp_lifetime,
+                        ty,
+                        span: expr.span,
+                        kind: ExprKind::Literal {
+                            literal: Literal::Value {
+                                value: tcx.mk_const(ty::Const {
+                                    val,
+                                    ty,
+                                }),
+                            },
+                        },
+                    };
+                    let source = match did {
+                        Some(did) => {
+                            let substs = Substs::identity_for_item(cx.tcx(), did);
+                            let base = mk_const(cx.tcx(), ConstVal::Unevaluated(did, substs));
+                            trace!("{:#?}", base);
+                            let offset = mk_const(
+                                cx.tcx(),
+                                ConstVal::Value(Value::ByVal(PrimVal::Bytes(offset as u128))),
+                            );
+                            let bin = ExprKind::Binary {
+                                op: BinOp::Add,
+                                lhs: base.to_ref(),
+                                rhs: offset.to_ref(),
+                            };
+                            Expr {
+                                temp_lifetime,
+                                ty,
+                                span: expr.span,
+                                kind: bin,
+                            }
+                        },
+                        None => mk_const(
+                            cx.tcx(),
+                            ConstVal::Value(Value::ByVal(PrimVal::Bytes(offset as u128))),
+                        ),
+                    };
+                    ExprKind::Cast { source: source.to_ref() }
+                } else {
+                    ExprKind::Cast { source: source.to_ref() }
```

Comment:
Can this line be deduplicated?

Reply:
the `source` binding is of different type

Reply:
What about the result of calling `.to_ref()`?